### DISCUSSION TITLE
[Suggestion] Speed update information

### DISCUSSION THREAD

**Discussion by moonspell-blip** [koda-dernet/Side-Step#43] [Ideas]
Hello,
thanks for your work !
I take a look at your code and it seems it's based on the native ace-step release from the first days.
im no longer active suporting ace-step, as the developer deleted mostly all my helpfull msg in discord and im not happy with the way the release was and how it's now erm fixed.

over the last weeks i fixed the 1 day release code, improved a lot.
i now testet your lora code, as i like the way it looks and works.
but i found some problems and maybe this information will help, if you have the time, to make it even better.

i postet in the early day's of the release a detailed guide in discord how to make lora training -71.6% faster.
but dev just ignore the info and actual its deleted, maybe not understand as the release code seems only vibe coded.
maybe you can use it now for your trainer:
[EXECUTIVE_SUMMARY.md](https://github.com/user-attachments/files/25879892/EXECUTIVE_SUMMARY.md)
[performance_comparison.md](https://github.com/user-attachments/files/25879789/performance_comparison.md)
and here all you need to implement:
[IMPLEMENTATION_GUIDE.md](https://github.com/user-attachments/files/25879896/IMPLEMENTATION_GUIDE.md)

│ BEFORE (Original Side-Step):                         
│ ├─ 1 training step: 3-4 Sekunden                   
│ ├─ 100 steps (1 Epoch): 5-7 Minuten                 
│ └─ 10 Epochs: 50-70 Minuten                          

│ AFTER (Mit allen 5 Optimierungen):                  
│ ├─ 1 training step: 0.8-1.2 Sekunden (-75%)    
│ ├─ 100 steps (1 Epoch): 1.5-2 Minuten           
│ └─ 10 Epochs: 15-20 Minuten                        

│ SPEEDUP FAKTOR: ~3-4x

its in german, you can translate to your home language or give it a ai you code with.
i used claude code to make this comparison, its a analyse of my lora training code vs your training code.
so you can see directly what to do, to get -71.6% speed too. its not much work, as i did the most wen i fix the ace-step 1day code.
what really was/is a mess...another topic.

as i do not host any repos here, or will in the future, i now use your trainer + my speed fix.
but wont upload it, as i do not want to have the stress with "issues" other people have with code that run fine for me.
no offence, im happy about everyone is doing open source, but im long ago out if this.
but i never stop sharing knowledge or stop helping others, so here im posting the info for you and your user.

fell free to ask questions if you have, happy coding and faster training times :)

**EDIT:**
i tested also the ace-step transcriber/captioner, 44gb wasted space on my ssd...
for me both were not usefull for the music styles i train, like hardstyle, tekk, phonk...and my homelanguage.
i spend like a week to find a better way to prepare the training data.

you should take a look at "music flamingo" from nvidia.
do not use the 16gb safetensors file, its like 30gb vram wen working and if using bits/bytes it became very slow.
like 5 min per song...

i make a little tool working like this:
mp3 -> wav
original song wav -> music-flamingo-hf-GGUF (via llama_cpp.exe not the lib) = all needed caption data, bpm, key etc
original song wav -> (HF becruily) mel_band_roformer_vocals -> original song voval stem wav -> music-flamingo-hf-GGUF (via llama_cpp.exe not the lib) = songtext lyrics with suno style tags
gemini via free api key -> create from music flamingo caption text, lora training caption
gemini via free api key -> create from songtext lyrics text, lora training songtext format

before i enter the lora keyword i want to activate the trained lora.

**result is a file like this and takes like 1 minute per 4-5 min song:**
caption: Dark, hypnotic Minimal Techno track defined by a stark, industrial-infused rhythmic evolution. Opening with a crisp, punchy four-on-the-floor kick and resonant synth bass, the piece establishes a cold, meditative foundation. Deep, authoritative male spoken-word vocals in German enter, delivering philosophical reflections with intimate, dry processing. As the arrangement progresses, sparse arpeggiated synth lines and subtle percussive clicks emerge, building a hypnotic tension through granular filter sweeps and layered noise textures. The track maintains a steady, relentless momentum, alternating between rhythmic intensity and transient harmonic shifts. The experience concludes with an atmospheric thinning of the electronic layers, allowing the resonant synth pads to drift into a final, hollow silence.
genre: Minimal Techno
tags: Minimal Techno, Industrial, Hypnotic, Introspective, Cold, Spoken-Word, Deep Male Vocal, Punchy Kick, Resonant Bass, Evolving Synth Textures, Precise Production, Dark Atmosphere, German Monologue, Rhythmic Loops, Berlin Style
bpm: 130
key: B Minor
signature: 4
custom_tag: TEKK
lyrics:
[Intro]
Ey, bleib mal stehen
Warte mal
Bleib einfach mal stehen
Schließ deine Augen
Und hör mir zu
Und hör mir zu
etc.

by using the gguf version, the lora data prep tool do not need more then 12gb vram.
so should work for mostly all users, i have 24gb but i like wen code does not need the max and is..lets say: clear coded by using multiple tools, if 1 tool did not do the job perfect.

maybe another idea for you, if you want to rework your i believe qwen-omni caption etc data prep function ;)

so, enough spam here. have fun :)

**Comment by thewwf** [comment]
Thanks for sharing your workflow, just tested music flamingo on the HF demo site, it transcribes audio very good. There's a comfyui node but it loads the safetensors files so it OOM's on 24gb vram cards. GGUF implementation of it would be a  great addition to Side-Step!

**Comment by moonspell-blip** [comment]
> Thanks for sharing your workflow, just tested music flamingo on the HF demo site, it transcribes audio very good. There's a comfyui node but it loads the safetensors files so it OOM's on 24gb vram cards. GGUF implementation of it would be a great addition to Side-Step!

you can use music flamingo as gguf via cli command.
no need to use comfyui etc. theres ways to use models before comfyui exists ;-)

just download from here: https://github.com/ggml-org/llama.cpp/releases
the two zip files for your cuda version and extract both in one folder.
let's say you use cuda 13, so download:
[Windows x64 (CUDA 13)](https://github.com/ggml-org/llama.cpp/releases/download/b8276/llama-b8276-bin-win-cuda-13.1-x64.zip) - [CUDA 13.1 DLLs](https://github.com/ggml-org/llama.cpp/releases/download/b8276/cudart-llama-bin-win-cuda-13.1-x64.zip)
(or the Linux/Mac Versions...)

you then use a cli command like this:
`llama_cpp\llama-mtmd-cli.exe ^
  --model .\models\music-flamingo-hf-GGUF\music-flamingo-hf.Q8_0.gguf ^
  --mmproj .\models\music-flamingo-hf-GGUF\music-flamingo-hf.mmproj-f16.gguf ^
  --audio .\temp\my_style_01.wav ^
  --prompt "<|im_start|>system\nYou are a helpful assistant specialized in music analysis.<|im_end|>\n<|im_start|>user\nWrite a rich caption that blends the technical details (genre, BPM, key, chords, mix) with how the song feels emotionally and dynamically as it unfolds.<|im_end|>\n<|im_start|>assistant\n" ^
  --n-predict 4096 ^
  --ctx-size 32768 ^
  --threads 8 ^
  --n-gpu-layers 999 ^
  --temp 0.3 ^
  --top-p 0.9`

result are detailed caption infos, you can feed gemini with and that you want them in the format side-step uses for training.
same goes with songtext, but other prompt. on the HF music flamingo page are some details what prompts used, take a look so you see what possible or write a prompt yourself for your needs.

the q8 model is like 12gb vram, if you need lower vram you can use q6 etc.

**edit:** use this gguf
https://huggingface.co/mradermacher/music-flamingo-hf-GGUF

**Comment by thewwf** [comment]
got it working, thanks for the suggestions!

**Comment by moonspell-blip** [comment]
> @moonspell-blip wir können in dein offenem Post weiter reden :)

@koda-dernet 
ah jetzt erst die notify gesehen und das du deutsch sprichst ;)
eben einen fehler in dem vergleich gefixt, das geht natürlich nicht 230% fixer...da habe ich beim posten gepennt.
das sind -71%.

![Image](https://github.com/user-attachments/assets/1fcf1c7a-8242-48e8-8543-8edf041e12f1)

![Image](https://github.com/user-attachments/assets/a06a08cd-c2bb-4769-8e0c-a4d337b124c2)

die bilder beziehen sich aber auf den day 1 release von ace-step, glaub du hast da schon in deinem trainer einige verbesserungen eingebaut betreffend dem speed. bin noch nicht dazu gekommen genauer hinzuschauen und mal was zu trainieren, aktuell echt viel in RL zu tun :(

wenn es bei mit wieder ruhiger ist, schau ich mir music flamingo noch mal an. das geht sicher auch mit der python lib, statt der cli und dann könnte ich dir den code posten. falls du das zum erstellen der caption/songtexte nutzen willst.
der free ac bei gemini hat 1.500 anfragen am tag, 5 in der minute. das reicht dicke.

**Comment by koda-dernet** [comment]
@moonspell-blip 

Danke dir für die Hinweise und die Doku, da sind schon ein paar interessante Punkte dabei

Ich würde den Vergleich zu Side-Step aber nicht 1:1 so übernehmen, weil ein Teil davon auf Annahmen basiert, die in der aktuellen Codebase so nicht mehr stimmen. Ich habe mir das angesehen: Side-Step hat z.B. bereits Warmup + SequentialLR, mehrere VRAM-/Training-Hebel wie Gradient Checkpointing, Encoder-Offloading und 8-bit Optimizer-State, und Music Flamingo ist bereits als Metadata-/Lyrics-Provider integriert, später als du es gesagt hast aber das war einem PR.

Deshalb würde ich den aktuellen Stand eher so einschätzen: Ein Teil deiner Hinweise sind echte Ideen, ein Teil ist aber eher gegen einen älteren oder unvollständigen Stand von Side-Step formuliert. Am interessantesten finde ich davon noch torch.compile als Benchmark-Thema und eventuell einen effizienteren Collate-/DataLoader-Pfad. Das sind Dinge, die man real messen kann.

Beim quantisierten DiT bzw. GGUF fürs eigentliche Training bin ich dagegen eher zurückhaltend. Für Inferenz oder Data-Prep kann das absolut sinnvoll sein. Aber beim LoRA-Training ist für mich die Kernfrage: Warum sollte ich auf einer quantisierten Approximation trainieren, wenn der Adapter am Ende eigentlich für das originale Base-Modell gedacht ist? Dann optimiere ich im Zweifel nicht nur auf Stil und Inhalt, sondern auch auf die veränderte Numerik bzw. Quantisierungseffekte des Backends.

Anders gesagt, GGUF kann für lokale Inferenz, Captioning oder Metadaten-Extraktion sehr interessant sein. Als allgemeine Verbesserung für das eigentliche Training sehe ich es aber nicht automatisch. Da würde ich eher bei den echten Gewichten bleiben und VRAM an anderer Stelle sparen.

Trotzdem danke für die Hinweise, vor allem die Punkte, die sich sauber benchmarken lassen

Diese Antwort ist vor allem, um zu fragen, ob es etwas Bestimmtes in diesem Codebase gibt, das du optimieren wolltest, oder ob ich deinen Beitrag vielleicht falsch verstanden habe.
